File: LeetCode/33.php
<?php
class Solution
{

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer
     */
    function search($nums, $target)
    {
        $count = count($nums);
        if ($nums[0] < $nums[$count - 1] && ($target < $nums[0] || $target > $nums[$count - 1])) return -1;
        $left = 0;
        $right = $count - 1;
        while ($left < $right) {
            $mid = ($left + $right) >> 1;
            if ($nums[$mid] == $target) return $mid;
            if ($nums[$mid] < $target) {
                if ($nums[$mid] <= $nums[$left]) {
                    if ($nums[$right] >= $target)
                        $left = $mid + 1;
                    else
                        $right = $mid - 1;
                } else {
                    $left = $mid + 1;
                }
            } else {
                if ($nums[$mid] >= $nums[$left]) {
                    if ($nums[$left] <= $target)
                        $right = $mid - 1;
                    else
                        $left = $mid + 1;
                } else {
                    $right = $mid - 1;
                }
            }
        }
        if ($nums[$left] == $target) return $left;
        return -1;
    }
}


$test = new Solution();
var_dump($test->search([1, 3], 3));

File: LeetCode/33.java
class Solution
{
    public int search(int[] nums, int target) {
        int length = nums.length;
        if (nums[0] < nums[length - 1] && (target < nums[0] || target > nums[length - 1]))
            return -1;
        int left = 0;
        int right = length - 1;

        while (left < right) {
            int mid = left + (right - left) >> 1;
            if (nums[mid] == target) return mid;
            if (nums[mid] < target) {
                if (nums[mid] <= nums[left] && nums[right] < target)
                    right = mid - 1;
                else
                    left = mid + 1;
            } else {
                if (nums[mid] >= nums[left] && nums[left] > target)
                    left = mid + 1;
                else
                    right = mid - 1;
            }
        }
        if (nums[left] == target)
            return left;
        return -1;
    }
}
